python/xoscar/collective/process_group.py

```diff
@@ -321,3 +330,314 @@ def broadcast(
+        recv_buf = convert_data_to_cp_array(recv_buf)
+        if self._is_world:
+            send_buf = cupy.concatenate(send_buf).reshape(self.world_size, -1)
+            stream = stream if stream != 0 else None
+            self._backend.scatter(send_buf, recv_buf, root, stream)
+        else:
+            if self.rank == root:
+                assert (
+                    len(send_buf) == self.world_size
+                ), "Scatter size must be equal to the size of group"
+                for peer in range(self.world_size):
+                    send_data = send_buf[peer]
+                    if peer == root:
+                        recv_buf[:] = send_data
+                    else:
+                        dtype = send_data.dtype
+                        self._backend.send(
+                            send_data.data.ptr,
+                            send_data.size,
+                            TypeMappingNCCL[dtype.type],
+                            peer,
+                            stream,
+                        )
+            else:
+                dtype = recv_buf.dtype
+                self._backend.recv(
+                    recv_buf.data.ptr,
+                    recv_buf.size,
+                    TypeMappingNCCL[dtype.type],
+                    root,
+                    stream,
+                )
+
+    def reduce_scatter(
+        self,
+        send_buf: Any,
+        recv_buf: Any,
+        recv_elems: List[int],
+        op: CollectiveReduceOp = CollectiveReduceOp.SUM,
+        stream: Optional[int] = 0,
+    ):
+        send_buf = convert_data_to_cp_array(send_buf)
+        recv_buf = convert_data_to_cp_array(recv_buf)
+        dtype = send_buf.dtype
+        if self._is_world:
+            stream = stream if stream != 0 else None
+            self._backend.reduce_scatter(
+                send_buf,
+                recv_buf,
+                recv_elems[self.rank],
+                ReduceOpMappingNCCLStr[op],
+                stream,
+            )
+        else:
+            self._backend.reduceScatter(
+                send_buf.data.ptr,
+                recv_buf.data.ptr,
+                recv_elems[self.rank],
+                TypeMappingNCCL[dtype.type],
+                ReduceOpMappingNCCL[op],
+                stream,
+            )
+
+    def alltoall(self, send_buf: Any, recv_buf: Any, stream: Optional[int] = 0):
+        assert (
+            self.world_size == send_buf.shape[0]
+        ), "The first dim of send data must be equal to world size."
+        assert (
+            recv_buf.size == send_buf.size
+        ), "The size of send data must be equal to the size of recv data."
+        send_buf = convert_data_to_cp_array(send_buf)
+        recv_buf = convert_data_to_cp_array(recv_buf)
+        dtype = send_buf.dtype
+        if self._is_world:
+            stream = stream if stream != 0 else None
+            self._backend.all_to_all(send_buf, recv_buf, stream)
+        else:
```

Comment:
Can we use https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.nccl.groupStart.html to accelerate these?

Reply:
Done